### ISSUE TITLE
Will await Task.Yield() be optimized away?

### ISSUE THREAD

**Issue by CeSun** [dotnet/runtime#70579]
I am trying to do this on winform and maui framework:

```c#
private async void button1_Click(object sender, EventArgs e)
{
    while(true)
    {
        await Task.Yield();
    }
}
```

My expectation is that the interface will not become unresponsive, but the result is the opposite of what I thought, the interface is unresponsive anymore.

But my friend has no problem on Xamarin for Android, as expected.
I suspect that Yield is optimized away.

**Comment by Clockwork-Muse** [comment]
... having an infinite loop seems like a strange thing to do in an event handler.  Why are you doing that, as opposed to something like just `await`ing whatever task you've submitted?  What's the underlying problem you're trying to solve?

Outside of that, the delay/yield might not be optimized out, it's just that it might be getting picked up by the UI thread again before other processing happens.

**Comment by tfenise** [comment]
I don't think `Task.Yield()` can be optimized away. For Winform, it is probably because `WindowsFormsSynchronizationContext.Post()` "posts" a message, and "posted" messages have higher priority than input messages.

In Winform, you can try `System.Windows.Forms.Application.DoEvents()` to keep responsive.
By the way, WPF offers `System.Windows.Threading.Dispatcher.Yield(DispatcherPriority)`, with which a low priority could be given.

**Comment by CeSun** [comment]
> ... having an infinite loop seems like a strange thing to do in an event handler.  Why are you doing that, as opposed to something like just `await`ing whatever task you've submitted?  What's the underlying problem you're trying to solve?
> 
> Outside of that, the delay/yield might not be optimized out, it's just that it might be getting picked up by the UI thread again before other processing happens.

For example, I have a large amount of data that needs to be updated to the user interface. If I update it directly, the interface may get stuck, because it is a long time and CPU intensive work. So in order to prevent the UI thread's ticks from getting stuck, I want to use the state machine mechanism of the co process to divide the long CPU intensive code into multiple segments and execute it in multiple ticks of the UI thread.

**Comment by CeSun** [comment]
> I don't think `Task.Yield()` can be optimized away. For Winform, it is probably because `WindowsFormsSynchronizationContext.Post()` "posts" a message, and "posted" messages have higher priority than input messages.
> 
> In Winform, you can try `System.Windows.Forms.Application.DoEvents()` to keep responsive.
> By the way, WPF offers `System.Windows.Threading.Dispatcher.Yield(DispatcherPriority)`, with which a low priority could be given.

I am not sure that the input is scheduled by the synchronization context.

The simplified model I understand is as follows:

```csharp
main()
{
   while(!NeedExit)
   {
      Input.Tick();
      UI.Tick();
      SyncContext.Tick(); 
   }
}
```

If I do not call await task.yield (), my CPU intensive code is executed in the button event, possibly in the input.Tick() is executed at one time.

If I call await task.yield(), my CPU intensive code will be split into multiple code segments and sent or post to the synchronization context, which will schedule execution.

**Comment by Clockwork-Muse** [comment]
> I want to use the state machine mechanism of the co process to divide the long CPU intensive code into multiple segments and execute it in multiple ticks of the UI thread.

No you don't.  And what you're currently doing [isn't guaranteed to work anyways](https://docs.microsoft.com/en-us/dotnet/api/system.threading.tasks.task.yield?view=net-6.0):

> The synchronization context that is present on a UI thread in most UI environments will often prioritize work posted to the context higher than input and rendering work. For this reason, do not rely on `await Task.Yield();` to keep a UI responsive.

(this was essentially what my first comment was assuming was happening, even before I started searching)

What you actually want to do is run the CPU-intensive work on another thread, so that the UI thread isn't blocked at all, probably by starting something via `Task.Run(...)`.  [This blog post series goes over your two options, and how to interact back with the UI thread when appropriate](https://blog.stephencleary.com/2013/05/taskrun-vs-backgroundworker-round-1.html)

**Comment by CeSun** [comment]
> > I want to use the state machine mechanism of the co process to divide the long CPU intensive code into multiple segments and execute it in multiple ticks of the UI thread.
> 
> No you don't.  And what you're currently doing [isn't guaranteed to work anyways](https://docs.microsoft.com/en-us/dotnet/api/system.threading.tasks.task.yield?view=net-6.0):
> 
> > The synchronization context that is present on a UI thread in most UI environments will often prioritize work posted to the context higher than input and rendering work. For this reason, do not rely on `await Task.Yield();` to keep a UI responsive.
> 
> (this was essentially what my first comment was assuming was happening, even before I started searching)
> 
> What you actually want to do is run the CPU-intensive work on another thread, so that the UI thread isn't blocked at all, probably by starting something via `Task.Run(...)`.  [This blog post series goes over your two options, and how to interact back with the UI thread when appropriate](https://blog.stephencleary.com/2013/05/taskrun-vs-backgroundworker-round-1.html)
> 
> 

What should I do if my CPU intensive work is to update the UI?

For example, I have 2000 data and want to put them all on the interface.

If it is placed in another thread, it will cause an exception.

**Comment by Clockwork-Muse** [comment]
.... it seems unlikely that just adding 2000 elements would be sufficient to lock the UI for a large amount of time, but rather that whatever you're doing to _gather_ the data is taking a large amount of time.  If it's your UI code, you may need to refactor something - but what you may need to change is better handled in general coding forums.  If it's the UI framework, reporting performance problems is appreciated, but we'd need a repro.

> If it is placed in another thread, it will cause an exception.

You're right about this.  See some of the later entries for examples.  
 - If updating the UI is actually fast, but gathering the data is slow, just `await`ing the results should be sufficient.
 - If updating the UI is for some reason slow, you'd probably best be served by creating some form of `Progress<T>` that you can post results (possibly batched) to.  This would increase the time it takes to completely all the updates, but should keep it otherwise responsive.

In either case, your next step is figuring out which more specific part of your code is slow, then asking for more targeted advice on someplace like StackOverflow or the C# discord.

Unless you can provide a specific repro for slow UI in general, there isn't anything actionable on our end.

**Comment by tfenise** [comment]
> For example, I have 2000 data and want to put them all on the interface.

How do you put them all on the interface? Typically one uses some sort of data binding, rather than creating 2000 controls for all pieces of data or putting all data directly into the content of a control.

**Comment by buyaa-n** [comment]
I don't think it is runtime/threading issue, consider @Clockwork-Muse's suggestion.

Transferring to MAUI for  more suggestions on how such UI should be loaded
